tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Integration tests for upgrading WordPress charm."""
+
+import asyncio
+import copy
+import io
+import json
+import logging
+import re
+import subprocess  # nosec
+import time
+
+import kubernetes
+import ops.model
+import PIL.Image
+import pytest
+import pytest_operator.plugin
+import requests
+from playwright.async_api import async_playwright
+
+from tests.integration.wordpress_client_for_test import WordpressClient
+
+logger = logging.getLogger()
+
+POST_CONTENT = """\
+WordPress powers more than 39% of the web — a figure that rises every day.
+Everything from simple websites, to blogs, to complex portals and enterprise
+websites, and even applications, are built with WordPress. WordPress combines
+simplicity for users and publishers with under-the-hood complexity for
+developers. This makes it flexible while still being easy-to-use."""
+
+POST_CONTENT = POST_CONTENT.replace("\n", " ")
+
+POST_TITLE = "WordPress Post #1"
+
+POST_COMMENT = "I am a comment."
+
+
+async def screenshot(url, path):
+    """Create a screenshot of a website."""
+    async with async_playwright() as playwright:
+        browser = await playwright.chromium.launch()
+        page = await browser.new_page()
+        await page.goto(url)
+        time.sleep(10)
+        await page.screenshot(full_page=True, path=path)
+        await browser.close()
+
+
+@pytest.fixture(scope="module", name="gen_upgrade_test_charm_config")
+def gen_upgrade_test_charm_config_fixture(ops_test, swift_config, kube_core_client):
+    """Create a function that generates charm config for upgrading tests."""
+    swift_url = swift_config["swift-url"]
+    swift_config = copy.copy(swift_config)
```

Comment:
is a shallow copy ok here? Otherwise there is also the `deepcopy` which copies not just the top level variables but does it recursively

Reply:
`swift_config` is a `Dict[str, str]`, that's why shallow copy is okay here.